horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java

```diff
@@ -570,17 +574,53 @@ Response addRunFromData(String start, String stop, String test,
       DatastoreResponse response = datastore.handleRun(data, metadata, testEntity.backendConfig, Optional.ofNullable(schemaUri), mapper);
 
       List<Integer> runIds = new ArrayList<>();
-      if (datastore.uploadType() == Datastore.UploadType.MUILTI && response.payload instanceof  ArrayNode){
-         response.payload.forEach(jsonNode -> {
-            runIds.add(getPersistRun(start, stop, test, owner, access, token, schemaUri, description, null, jsonNode, testEntity));
-         });
+      String responseString;
+      if (datastore.uploadType() == Datastore.UploadType.MUILTI
+              && response.payload instanceof ArrayNode ){
+
+         //if we return more than 10 results, offload to async queue to process - this might take a LOOONG time
+         if ( response.payload.size() > 10 ) {
+            response.payload.forEach(jsonNode -> {
+               mediator.queueRunUpload(start, stop, test, owner, access, token, schemaUri, description, null, jsonNode, testEntity);
+            });
+         } else { //process synchronously
+            response.payload.forEach(jsonNode -> {
+               runIds.add(getPersistRun(start, stop, test, owner, access, token, schemaUri, description, metadata, response.payload, testEntity));
+            });
+         }
       } else {
          runIds.add(getPersistRun(start, stop, test, owner, access, token, schemaUri, description, metadata, response.payload, testEntity));
       }
-//      return Response.status(Response.Status.OK).entity(String.valueOf(runId)).header(HttpHeaders.LOCATION, "/run/" + runId).build();
-      String reponseString = String.valueOf(runIds.stream().map(val -> Integer.toString(val)).collect(Collectors.joining(", ")));
-      return Response.status(Response.Status.OK).entity(reponseString).build();
+
+      responseString = runIds.size() > 0 ?
+              String.valueOf(runIds.stream().map(val -> Integer.toString(val)).collect(Collectors.joining(", ")))
+              : "More than 10 runs uploaded, processing asynchronously";
+
+      return Response.status(Response.Status.OK).entity(responseString).build();
    }
+   @Transactional
+   public void persistRun(ServiceMediator.RunUpload runUpload) {
```

Comment:
Otherwise we would expose `ServiceMediator.RunUpload` outside its scope.
```suggestion
   void persistRun(ServiceMediator.RunUpload runUpload) {
```

Reply:
Hmm... `ServiceMediator.RunUpload` is already outside of scope here, because this method couples RunServiceImpl with ServiceMediator... which I don't think is a great pattern.  

tbh, i don't know how much of the codebase is tightly coupled like this

Reply:
>ServiceMediator.RunUpload is already outside of scope here

Well, from java "Access modifier" point of view if we remove the `public` from `persitRun` method it shouldn't be out of scope here but yeah I agree that this is going to thigthly couple the RunServiceImpl with ServiceMediator even if I don't think this is an issue here